### ISSUE TITLE
tracing-subscriber 0.3.1 does not respect RUST_LOG

### ISSUE THREAD

**Issue by theduke** [tokio-rs/tracing#1697]
### Version

tracing-subscriber 0.3.1

### Platform

Linux

### Description

tracing-subscriber seems to incorrectly handle the RUST_LOG env var.

After `tracing_subscriber::fmt::init()` I only get `INFO` and above events,  regardless of the `RUST_LOG` env var (RUST_LOG=trace, RUST_LOG=crate=trace, ...)

Downgrading to `0.2` fixes this.

(I checked the release notes for changed behaviour, but it seems like there aren't any relevant alterations)

**Comment by davidbarsky** [comment]
I’ll update the changelog later, but the `EnvFilter` functionality is now behind an off-by-default feature flag called `env-filter`. You’ll need to add that to your Cargo.toml.

(The documentation is thankfully updated: https://docs.rs/tracing-subscriber/0.3.1/tracing_subscriber/struct.EnvFilter.html)

**Comment by hawkw** [comment]
> I’ll update the changelog later, but the `EnvFilter` functionality is now behind an off-by-default feature flag called `env-filter`. You’ll need to add that to your Cargo.toml

The changelog does mention the feature flagging change: https://github.com/tokio-rs/tracing/blob/v0.1.x/tracing-subscriber/CHANGELOG.md#breaking-changes but it may not be obvious what _behavior_ this changes, so we might want to clarify that changelog entry.

I think the most important docs improvement, though, might be to change the documentation for [methods like `fmt::init`](https://docs.rs/tracing-subscriber/0.3.1/tracing_subscriber/fmt/fn.init.html), that include an `EnvFilter` in the configured subscriber, to explicitly state that the feature flag is necessary for `RUST_LOG` filtering.

**Comment by rcoh** [comment]
perhaps we could log a warning if `RUST_LOG` is set but the feature isn't enabled in `tracing_subscriber::fmt::init()`? I suspect this will save a lot of people a lot of debugging time.

**Comment by dzamlo** [comment]
Despite enabling the `env-filter`,  using `tracing_subscriber::fmt::init()` is not enough.

I have to manually add the filter:
```
tracing_subscriber::fmt()
        .with_env_filter(EnvFilter::from_default_env())
        .with_writer(non_blocking)
        .init();
```

Is it corect that this is not the expected behaviour?

**Comment by Firstyear** [comment]
I can reproduce this as well:

```
[package]
name = "tracing_demo"
version = "0.1.0"
edition = "2021"

[dependencies]
tracing = "0.1"
tracing-subscriber = { version = "0.3", features = ["env-filter", "std", "fmt"] }
```

```
#[macro_use]
extern crate tracing;

fn main() {
    tracing_subscriber::fmt().init();

    debug!("This is a debug message");

    trace!("This is a trace message");
}
```

Using RUST_LOG=trace or debug both will not show the above messages. 

```
tracing-subscriber v0.3.5
tracing v0.1.29
rustc 1.57.0 (f1edd0429 2021-11-29)
```

**Comment by kafji** [comment]
@Firstyear that's because `tracing_subscriber::fmt().init()` is not same as `tracing_subscriber::fmt::init()`.

- `tracing_subscriber::fmt::init()` => initialize subscriber with opinionated configurations i.e. with `with_env_filter(EnvFilter::from_default_env())`
- `tracing_subscriber::fmt().init()` => create subscriber builder (with empty configurations) and then initialize it as global subscriber

It is confusing, https://github.com/tokio-rs/tracing/issues/1329

tl;dr
- add `with_env_filter(EnvFilter::from_default_env())`
- or use `fmt::init()`

**Comment by Firstyear** [comment]
Yes, I think the docs also suggested at one point `.init()` as well, as I remember copy-pasting this line.

**Comment by davidbarsky** [comment]
Gonna close this since this was fixed in [tracing-subscriber 0.3.5](https://github.com/tokio-rs/tracing/releases/tag/tracing-subscriber-0.3.5) with https://github.com/tokio-rs/tracing/pull/1781.
